### PR TITLE
FIR checker: report EXPECTED_CONDITION and ILLEGAL_DECLARATION_IN_WHEN_SUBJECT

### PR DESCRIPTION
**Pull request by tgeng**

### CODE REVIEW

**Review by punzki — APPROVED** [review]

File: compiler/testData/diagnostics/tests/when/withSubjectVariable/unsupportedVariableDeclarationsInWhenSubject.fir.kt [hunk 1]
```diff
@@ -12,7 +12,7 @@ fun testSimpleValInWhenSubject() {
 
 fun testValWithoutInitializerWhenSubject() {
     when (val y: Any) {
-        is String -> <!UNINITIALIZED_VARIABLE!>y<!>.<!UNRESOLVED_REFERENCE!>length<!>
+        <!EXPECTED_CONDITION!>is String<!> -> <!UNINITIALIZED_VARIABLE!>y<!>.<!UNRESOLVED_REFERENCE!>length<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/when/withSubjectVariable/unsupportedVariableDeclarationsInWhenSubject.fir.kt:15]: Hm, the counterpart in FE1.0 rather reports an error on `y` above: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/when/withSubjectVariable/unsupportedVariableDeclarationsInWhenSubject.kt#L21 Can we avoid this?

Comment (tgeng) [reply compiler/testData/diagnostics/tests/when/withSubjectVariable/unsupportedVariableDeclarationsInWhenSubject.fir.kt:15]: Thanks! Let me suppress the error in this case.

**Review by jsjeon — APPROVED** [review]
Nice suppression! :)

**Comment by mglukhikh** [comment]
Merged manually. However I dropped NotRootCause https://github.com/JetBrains/kotlin/commit/bf40c07cc372fb8467f5a9dceb099bea515d2250 as it looks artificial for me. If we need it for some other diagnostics, we can revert this my commit.

**Pull request closed without merging**